keras/callbacks.py

```diff
@@ -858,44 +858,30 @@ def is_indexed_slices(grad):
 
             projector.visualize_embeddings(self.writer, config)
 
+    def _fetch_callback(self, summary):
+        self.writer.add_summary(summary, self._epoch)
+
+    def on_epoch_begin(self, epoch, logs=None):
+        if not self.params['do_validation'] and self.histogram_freq:
+            raise ValueError('If printing histograms, validation_data must be '
+                             'provided.')
+        if self.histogram_freq and epoch % self.histogram_freq == 0:
+            self._epoch = epoch
+            if self.merged not in self.model.test_function.fetches:
+                self.model.test_function.fetches.append(self.merged)
+                self.model.test_function.fetch_callbacks[self.merged] = self._fetch_callback
```

Comment:
Why use a callback system here?

Why not:

```python
summary = self.model.test_function.fetch_values[self.merged]
self.writer.add_summary(summary, self._epoch)
```

and in `Function`:

```python
self.fetch_values = dict(zip(self._fetches, updated[-len(self._fetches):]))
```

Or something like that. Wouldn't that be more general and easier to follow?

Reply:
What would trigger the write each batch in that case?

Reply:
I  believe the sequence of actions would be strictly equivalent to the current formulation? No?

Reply:
Using an array for each key in fetch_values would be an option.
In that case reacting to fetch_values in between the epoch callbacks wouldn't be possible.

Reply:
Maybe you could add the missing hooks (for example `on_valid/test_batch_start`, `on_valid/test_batch_end`), @hermansje. This would make things a little cleaner.

Reply:
For the summary names, I now applied the solution from Tensorflow.

Reply:
About the callbacks, I agree extra hooks would be a bit more general and familiar to use, but it would complicate the code, and cause API changes in `Callback` and in the `Model` functions.

It would add 4 hooks in `Callback` that have to be called in multiple places (generator or not + validation generator or not) that would also need access to the callbacks.
So the callbacks would have to be passed around and added as parameter of the functions in `Model`.

Is that impact justifiable for the (rather advanced / currently very limited) use of this feature?

A variation on the current implementation could be to use `defaultdict(set)` for the callbacks [like this](https://github.com/hermansje/keras/commit/57d67bf051efd5ac25f516b28c143564ba0cc7d2), making it more general (multiple callbacks per tensor without overwriting or creating a new tensor using an `id` op).